### PR TITLE
Remove unused traits

### PR DESCRIPTION
**Pull request by dhollman**
Removed more dead code related to metaprogramming facilities that I don't want to maintain (and that we don't need to maintain any more).

Removed:
* `if_`
* `is_label`
* `TypeList`
* `ReverseTypeList`

### CODE REVIEW

**Comment by dalg24** [comment]
Retest this please

**Review by masterleinad — APPROVED** [review]

**Review by dalg24 — DISMISSED** [review]
Didn't find any match in Trilinos nor Kokkos-Kernels

File: core/src/impl/Kokkos_Traits.hpp [hunk 1]
```diff
@@ -243,31 +243,10 @@ struct if_c<true, void, FalseType> {
...
-
-template <int N>
-struct is_label<char[N]> : public std::true_type {};
-
-template <>
-struct is_label<const std::string> : public std::true_type {};
-
-template <>
-struct is_label<std::string> : public std::true_type {};
+struct is_label : std::is_convertible<T, std::string> {};
```
Comment (dalg24) [inline core/src/impl/Kokkos_Traits.hpp:249]: I would prefer
```suggestion
struct is_label : std::is_convertible<T, std::string>::type {};
```

**Comment by Rombur** [comment]
Retest this please

**Review by dalg24 — DISMISSED** [review]
Please undo removal of `if_`

File: core/src/impl/Kokkos_Traits.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/impl/Kokkos_Traits.hpp:249]: I agree with it being its own PR.

I'm also wondering if under C++17 the test should be if it is convertible to `string_view`.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_Traits.hpp:249]: I just grepped for `is_label` in Trilinos develop branch and it is only used in this file.

**Pull request merged**